### ISSUE TITLE
Separate core metrics from user metrics pipeline

### ISSUE THREAD

**Issue by dashpole** [kubernetes/kubernetes#39102]
This issue is intended to foster discussion around the following proposal: [#252](https://github.com/kubernetes/community/pull/252)

TLDR: The kubelet doesn't use many of the metrics it collects.  We should move towards a structure in which the kubelet collects only metrics it uses ("Core Metrics"), and creates a structure in which users/third parties can collect metrics they want.  The most important aspect of this proposal are the Core Metrics that are collected.

Do any other k8s components (outside the kubelet) use the summary api?  In that case we would want to include whatever they use in the core metrics API.
A: Horizontal Pod Autoscalar requires pod-level and node-level cpu and memory metrics.  Scheduler requires pod-level, container-level and node-level CPU, Memory, and Disk metrics.

Should this API be versioned, or unversioned?  It would have to be versioned if another component relies on the summary API, which I am not sure of.
A: Versioned

Are there any metrics that we will need in the Q1 timeframe that should be included?
A: HPA and scheduler metrics include metrics to be used in the near future.

**Comment by dashpole** [comment]
cc: @kubernetes/sig-scheduling, @kubernetes/sig-instrumentation

**Comment by dchen1107** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/27097#issuecomment-237422351

**Comment by dashpole** [comment]
scheduling: [#27097 (comment)](https://github.com/kubernetes/kubernetes/issues/27097#issuecomment-237962464)

**Comment by dashpole** [comment]
Possibly a better way to approach this proposal, is based on which metrics currently included in the [Summary API](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/stats/types.go) should NOT be included in the core metrics API.
The current proposal for core metrics (based on what the kubelet uses) does NOT include the following stats:
Timestamp data for any metrics
Identifying information other than podUID
System Container Stats
Node-level CPU Stats
Node-level Network Stats
Container-level CPU Stats
Container-level user-defined metrics.

@davidopp For Scheduling, what kind of cpu metrics would we need?  Do we need node-level, pod-level, container-level, or all three?

**Comment by piosz** [comment]
cc: @kubernetes/sig-scheduling-misc @kubernetes/sig-instrumentation-misc

**Comment by davidopp** [comment]
@dashpole I've written up a [short doc](https://docs.google.com/a/google.com/document/d/1gU6ji4pA7YtFF3AYRH4rQ0YFB2y-3Pu9Daq2mBNPrZI/edit?usp=sharing) to answer your question. (I started writing it in Github but it was getting very long.) The tl;dr is that I believe we need per-pod resource usage, and either per-node usage or the sum of the usages of the non-pod resource consumers (like Docker daemon).

**Comment by derekwaynecarr** [comment]
@dashpole -- the api needs to be versioned.  i would like us to clean up the kubelet apiserver portion of the code as well in this work.

**Comment by DirectXMan12** [comment]
> Do any other k8s components (outside the kubelet) use the summary api? In that case we would want to include whatever they use in the core metrics API.

Heapster (currently) and metrics-server (upcoming) use it to get metrics for the Horizontal Pod Autoscaler controller.  We'll need CPU and memory for that.

IIRC, wasn't plan was to eventually drive the scheduler off of the resource metrics API (which is served by metrics-server)?

**Comment by DirectXMan12** [comment]
cc @kubernetes/sig-autoscaling-misc

**Comment by DirectXMan12** [comment]
I think it needs to be clearer as to how this fits into the general monitoring vision.  As far as I see it, this would replace the summary API as the source of information for metrics-server, and the rest of the kubernetes components would consume from metrics-server (via the resource metrics API), according to the existing doc.

**Comment by davidopp** [comment]
> IIRC, wasn't plan was to eventually drive the scheduler off of the resource metrics API (which is served by metrics-server)?

Yes, I was assuming this is still the plan.

Note that we don't actually use any usage metrics in the scheduler right now. We were actually waiting for the metrics API (and having time).

**Comment by davidopp** [comment]
> The tl;dr is that for scheduling I believe we need per-pod resource usage, and either per-node usage or the sum of the usages of the non-pod resource consumers (like Docker daemon).

@vishh pointed out that the first part of this (per-pod resource usage) is wrong; we actually need per-container. This is because request and limit are set per-container. So vertical autoscaling needs to know the per-container usage, since the knob it tunes is per-container. We could say vertical autoscaling (and humans who are trying to figure out how to set request/limit for their containers) needs to use the monitoring pipeline rather than the core metrics pipeline but there are disadvantages of doing that.

**Comment by xiang90** [comment]
@DirectXMan12 

> metrics-server (upcoming)

Do you have a design doc or a short introduction of this project?

**Comment by piosz** [comment]
@xiang90 metrics-server is somewhat described in [monitoring architecture](https://github.com/kubernetes/kubernetes/blob/master/docs/design/monitoring_architecture.md). It's more or less trimmed down [Heapster](https://github.com/kubernetes/heapster) without sinks, exporting only one well defined [Master Metrics API](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-metrics-api.md), implemented by reusing genericapiserver library which will allow to make it discoverable from discovery summarizer.

**Comment by dashpole** [comment]
Proposal updated to include:

- Versioned API

- HPA and Scheduler Resource requirements

**Comment by dashpole** [comment]
cc: @kubernetes/sig-node-misc

**Comment by derekwaynecarr** [comment]
@DirectXMan12 that is my understanding as well, but we lack any of the machinery in the kubelet to support stable versioned apis (hence why everything is staying at v1alpha1)

**Comment by dashpole** [comment]
The new proposal (in kubernetes/community) can be [found here](https://github.com/kubernetes/community/pull/252)

**Comment by dashpole** [comment]
Closing this for now, as the proposal was generally approved, and there is no current need for this issue.
